### DISCUSSION TITLE
Doubt on SRP

### DISCUSSION THREAD

**Discussion by rohithanaydy** [openthread/openthread#7553] [Q&A] [answered]
I am trying to implement SRP protocol in my Thread Network consist of
 1 Border Router(RCP as Leader),1 Router and 1 Child based on [https://openthread.io/codelabs/openthread-border-router#4](url)
I got the result and obtained the services on the **Service Browser** app. I have doubt in the service, Can we use only standard services  like _test1._udp, _test2._udp and _ipps._tcp ?
When  I used _temp._sensor this type of service it is not registered.  How can we use our own services ?
(I am using SRP to know the Ipv6 of a node in the Thread Network ,which is temperature sensor) ?
The number of services of a single SRP client can register  is limited to 2 ?

Thank You
Rohith R

**Comment by jwhui** [comment, marked as answer]
> When I used _temp._sensor this type of service it is not registered. How can we use our own services ? (I am using SRP to know the Ipv6 of a node in the Thread Network ,which is temperature sensor) ?

Your service name is not structured properly. [RFC 6763 Section 4.1.2](https://datatracker.ietf.org/doc/html/rfc6763#section-4.1.2) states:

```
   The <Service> portion of the Service Instance Name consists of a pair
   of DNS labels, following the convention already established for SRV
   records [RFC2782].  The first label of the pair is an underscore
   character followed by the Service Name [RFC6335].  The Service Name
   identifies what the service does and what application protocol it
   uses to do it.  The second label is either "_tcp" (for application
   protocols that run over TCP) or "_udp" (for all others).  For more
   details, see Section 7, "Service Names".
```

Try something like `_temp._udp`.

> The number of services of a single SRP client can register is limited to 2 ?

You can configure the number of services SRP client supports using [`OPENTHREAD_CONFIG_SRP_CLIENT_BUFFERS_MAX_SERVICES`](https://github.com/openthread/openthread/blob/5eb973ce00d3eec5898d6cc63fd4a20f9e914111/src/core/config/srp_client.h#L307).

**Comment by abtink** [comment]
> The number of services of a single SRP client can register is limited to 2 ?

SRP client implementation in OpenThread does not add a limit on the number of services that can be registered.

It expects the service entry strutures [`otSrpClientService`](https://github.com/openthread/openthread/blob/5eb973ce00d3eec5898d6cc63fd4a20f9e914111/include/openthread/srp_client.h#L83-L112) to be provided by the application code (user of OT SRP APIs). So it is up the application code to determine how many services it wants and allocate the proper number of entries. This makes its design flexible under different use-cases.

There is also another module [`otSrpClientBuffers`](https://github.com/openthread/openthread/blob/5eb973ce00d3eec5898d6cc63fd4a20f9e914111/include/openthread/srp_client_buffers.h)  which can be [enabled/disabled](https://github.com/openthread/openthread/blob/5eb973ce00d3eec5898d6cc63fd4a20f9e914111/src/core/config/srp_client.h#L288-L296) independently of SRP client itself. This module provide a pool of service entries for application code to use. This is intended to help simpler application code which do not want to allocate and manage such buffers. This is where the [OPENTHREAD_CONFIG_SRP_CLIENT_BUFFERS_MAX_SERVICES](https://github.com/openthread/openthread/blob/5eb973ce00d3eec5898d6cc63fd4a20f9e914111/src/core/config/srp_client.h#L307) is used and defined.

I suggest based on your application to evaluate whether you want to enable and use`otSrpClientBuffers` (then you may need to properly configure it) or have your own allocation mechanism in the application layer code.
